Switched to single test framework per TestModule

lefou opened the pull request:
Introduced new methods and targets where appropriate.
Deprecated all methods that support multiple frameworks.

See also discussion in https://github.com/com-lihaoyi/mill/issues/170

lolgab commented on scalajslib/src/ScalaJSModule.scala:
```diff
@@ -218,13 +218,13 @@ trait TestScalaJSModule extends ScalaJSModule with TestModule {
     val (close, framework) = mill.scalajslib.ScalaJSWorkerApi.scalaJSWorker().getFramework(
       toolsClasspath().map(_.path),
       jsEnvConfig(),
-      testFrameworks().head,
```
Beautiful!

lolgab commented on scalalib/src/TestModule.scala:
```diff
@@ -0,0 +1,215 @@
+  trait TestNg extends TestModule {
+    override def ivyDeps: T[Agg[Dep]] = T {
+      super.ivyDeps() ++ Agg(
+        ivy"com.lihaoyi:mill-contrib-testng_2.13:${mill.BuildInfo.millVersion}")
```
https://github.com/com-lihaoyi/mill/pull/1268/files#diff-aa5cde7401023fc2ba3efedb50e338ac1643922517017182878165d8b05af73cR24
Here you have `mill-contrib-testng` without `_2.13` suffix. Why?

lefou replied:
I think the one in caffeine tests is unused, as the test framework in use is JUnit. We need to review caffeine tests further, to make sure the setup is correct. I remember that I hat lots of issues with caffeine builds when I added integration tests for `mill-cobtrib-testng`.

lefou replied:
FTR, https://github.com/com-lihaoyi/mill/pull/566#issuecomment-470440705

lolgab reviewed (changes requested):
It seems strange to have `mill-contrib-testng` artifact with an hardcoded `_2.13`.
I guess it's to be used with Java projects.
If you mix it with other Scala dependencies not on Scala 2.13 it silently breaks the binary compatibility.
Maybe we should rewrite that module in pure java if possible?
Other than this it looks good to me!

lefou commented:
> It seems strange to have `mill-contrib-testng` artifact with an hardcoded `_2.13`.
> I guess it's to be used with Java projects.
> If you mix it with other Scala dependencies not on Scala 2.13 it silently breaks the binary compatibility.
> Maybe we should rewrite that module in pure java if possible?
> Other than this it looks good to me!

Oh, our testng framework is already a pure Java implementation. I think the Scala binary suffix is an oversight due to current consistent packaging setup.

lolgab commented:
> > It seems strange to have `mill-contrib-testng` artifact with an hardcoded `_2.13`.
> > I guess it's to be used with Java projects.
> > If you mix it with other Scala dependencies not on Scala 2.13 it silently breaks the binary compatibility.
> > Maybe we should rewrite that module in pure java if possible?
> > Other than this it looks good to me!
> 
> Oh, our testng framework is already a pure Java implementation. I think the Scala binary suffix is an oversight due to current consistent packaging setup.

I think it was a side effect of [this change](https://github.com/com-lihaoyi/mill/commit/bb8c47dd7b3136bc081928a68c16e8a61ed46eb4#diff-c6f64a5895d9f518051cf83f8557f492ec6f5683a34c2914a46a894001e62726L364-R364)
If you look at Maven Central at some point ( a few commits after this ) the artifact was renamed with the extra `_2.13`. This because `MillModule` depends on Scala where `MillPublishModule` depends on Java only.

lolgab commented:
From my side the issue with TestNG can be fixed later with a separate PR.
Just add a `TODO` where the dependency is added with the `_2.13` and create a companion issue to link in the `TODO`

lolgab commented on build.sc:
```diff
@@ -444,6 +444,8 @@ object scalajslib extends MillModule {
 object contrib extends MillModule {
   object testng extends MillModule{
+    // pure Java implementation
+    override def artifactSuffix: T[String] = ""
```
I'm not sure this works. The classpath still depends on Scala 2.13.
For this to work it should not depend on `MillModule`. The scope is big enough to be fixed separately.

```
lorenzo@MacBook-Pro-di-Lorenzo ~/s/mill (testframework) [1|1]> ./mill show contrib.testng.compileClasspath | grep "_2.13"
[1/1] show 
[1/1] show > [55/232] de.tobiasroeser.mill.vcs.version.VcsVersion.vcsState 
[1/1] show > [221/232] contrib.testng.resources 
    "qref:a7f1b3be:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/scalafmt-dynamic_2.13/2.7.5/scalafmt-dynamic_2.13-2.7.5.jar",
    "qref:9a3798c5:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite_2.13.3/2.3.8-46-37b110d0/ammonite_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:1d604d75:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/trees_2.13/4.4.13/trees_2.13-4.4.13.jar",
    "qref:b7e76f91:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier_2.13/2.0.16/coursier_2.13-2.0.16.jar",
    "qref:df997342:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/eed3si9n/jarjarabrams/jarjar-abrams-core_2.13/0.3.1/jarjar-abrams-core_2.13-0.3.1.jar",
    "qref:17c63455:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/sourcecode_2.13/0.2.5/sourcecode_2.13-0.2.5.jar",
    "qref:321465a3:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/os-lib_2.13/0.7.4/os-lib_2.13-0.7.4.jar",
    "qref:c52b33d6:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle_2.13/1.3.11/upickle_2.13-1.3.11.jar",
    "qref:3d0a8eab:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-terminal_2.13/2.3.8-46-37b110d0/ammonite-terminal_2.13-2.3.8-46-37b110d0.jar",
    "qref:179efc28:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-ops_2.13/2.3.8-46-37b110d0/ammonite-ops_2.13-2.3.8-46-37b110d0.jar",
    "qref:732dd1cf:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-util_2.13/2.3.8-46-37b110d0/ammonite-util_2.13-2.3.8-46-37b110d0.jar",
    "qref:87a0917d:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-runtime_2.13.3/2.3.8-46-37b110d0/ammonite-runtime_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:77110e9d:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-interp-api_2.13.3/2.3.8-46-37b110d0/ammonite-interp-api_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:5fd5c3d2:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-repl-api_2.13.3/2.3.8-46-37b110d0/ammonite-repl-api_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:133aecbb:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-interp_2.13.3/2.3.8-46-37b110d0/ammonite-interp_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:f201cdd3:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-repl_2.13.3/2.3.8-46-37b110d0/ammonite-repl_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:b90a54fe:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-compiler_2.13.3/2.3.8-46-37b110d0/ammonite-compiler_2.13.3-2.3.8-46-37b110d0.jar",
    "qref:2307792a:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/common_2.13/4.4.13/common_2.13-4.4.13.jar",
    "qref:11602913:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/thesamet/scalapb/scalapb-runtime_2.13/0.11.0/scalapb-runtime_2.13-0.11.0.jar",
    "qref:e2edc2ac:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/fastparse-v2_2.13/2.3.1/fastparse-v2_2.13-2.3.1.jar",
    "qref:636c1e09:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-core_2.13/2.0.16/coursier-core_2.13-2.0.16.jar",
    "qref:de846188:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-cache_2.13/2.0.16/coursier-cache_2.13-2.0.16.jar",
    "qref:293f3f82:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/github/alexarchambault/argonaut-shapeless_6.2_2.13/1.2.0/argonaut-shapeless_6.2_2.13-1.2.0.jar",
    "qref:f291d899:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/geny_2.13/0.6.8/geny_2.13-0.6.8.jar",
    "qref:910c9083:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ujson_2.13/1.3.11/ujson_2.13-1.3.11.jar",
    "qref:da930705:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upack_2.13/1.3.11/upack_2.13-1.3.11.jar",
    "qref:3c69c77d:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle-implicits_2.13/1.3.11/upickle-implicits_2.13-1.3.11.jar",
    "qref:25f36c9b:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/fansi_2.13/0.2.9/fansi_2.13-0.2.9.jar",
    "qref:488979ed:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-collection-compat_2.13/2.4.2/scala-collection-compat_2.13-2.4.2.jar",
    "qref:e003f5ab:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/pprint_2.13/0.6.0/pprint_2.13-0.6.0.jar",
    "qref:a8ab3e64:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/requests_2.13/0.6.5/requests_2.13-0.6.5.jar",
    "qref:ff528205:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/mainargs_2.13/0.1.4/mainargs_2.13-0.1.4.jar",
    "qref:cdef4387:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ammonite-compiler-interface_2.13/2.3.8-46-37b110d0/ammonite-compiler-interface_2.13-2.3.8-46-37b110d0.jar",
    "qref:3ea30874:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.13/2.0.0-M3/scala-xml_2.13-2.0.0-M3.jar",
    "qref:5fb9a714:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/scalaparse_2.13/2.3.0/scalaparse_2.13-2.3.0.jar",
    "qref:1544d6b7:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/thesamet/scalapb/lenses_2.13/0.11.0/lenses_2.13-0.11.0.jar",
    "qref:4e66b320:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-util_2.13/2.0.16/coursier-util_2.13-2.0.16.jar",
    "qref:752fa5e0:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/argonaut/argonaut_2.13/6.2.5/argonaut_2.13-6.2.5.jar",
    "qref:5d4e7ef4:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/chuusai/shapeless_2.13/2.3.3/shapeless_2.13-2.3.3.jar",
    "qref:36160659:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle-core_2.13/1.3.11/upickle-core_2.13-1.3.11.jar",
    "qref:b833763b:/Users/lorenzo/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/fastparse_2.13/2.3.0/fastparse_2.13-2.3.0.jar",
lorenzo@MacBook-Pro-di-Lorenzo ~/s/mill (testframework)> 
```

lefou replied:
Hm, this doesn't look right.

Here's mine:

```
14:50:01 ⌁ θ62° lefou:~/work/opensource/mill testframework(+7/-6)+* 130 ± mill show contrib.testng.compileClasspath
[1/1] show 
[1/1] show > [2/13] contrib.testng.resources 
[
    "ref:c984eca8:/home/lefou/work/opensource/mill/contrib/testng/resources",
    "qref:1e75ab0b:/home/lefou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.3/scala-library-2.13.3.jar",
    "qref:8c6af64e:/home/lefou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/test-interface/1.0/test-interface-1.0.jar",
    "qref:a12b3706:/home/lefou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/testng/testng/7.4.0/testng-7.4.0.jar",
    "qref:f37a4086:/home/lefou/.cache/coursier/v1/https/repo1.maven.org/maven2/com/beust/jcommander/1.78/jcommander-1.78.jar",
    "qref:1d8bf44c:/home/lefou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/webjars/jquery/3.5.1/jquery-3.5.1.jar"
]
```

I'd like to keep it simple, and we definitely want the test setup from `MillModule` which uses Scala + Mill + TestEvaluator.

We only have this dependency tree:

```
 14:39:53 ⌁ θ61° lefou:~/work/opensource/mill testframework(+7/-6)+* ± mill contrib.testng.ivyDepsTree
[6/6] contrib.testng.ivyDepsTree 
├─ org.scala-sbt:test-interface:1.0
└─ org.testng:testng:7.4.0
   ├─ com.beust:jcommander:1.78
   └─ org.webjars:jquery:3.5.1
14:48:18 ⌁ θ60° lefou:~/work/opensource/mill testframework(+7/-6)+* 1 ± mill contrib.testng.ivyDepsTree --withRuntime true
[7/7] contrib.testng.ivyDepsTree 
├─ org.scala-sbt:test-interface:1.0
└─ org.testng:testng:7.4.0
   ├─ com.beust:jcommander:1.78
   └─ org.webjars:jquery:3.5.1
```

And also the published pom has no scala dependencies:

```xml
    <dependencies>
        <dependency>
            <groupId>org.scala-sbt</groupId>
            <artifactId>test-interface</artifactId>
            <version>1.0</version>
        </dependency>
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.4.0</version>
        </dependency>
    </dependencies>
`` `

lefou commented:
> > > It seems strange to have `mill-contrib-testng` artifact with an hardcoded `_2.13`.
> > > I guess it's to be used with Java projects.
> > > If you mix it with other Scala dependencies not on Scala 2.13 it silently breaks the binary compatibility.
> > > Maybe we should rewrite that module in pure java if possible?
> > > Other than this it looks good to me!
> > 
> > 
> > Oh, our testng framework is already a pure Java implementation. I think the Scala binary suffix is an oversight due to current consistent packaging setup.
> 
> I think it was a side effect of [this change](https://github.com/com-lihaoyi/mill/commit/bb8c47dd7b3136bc081928a68c16e8a61ed46eb4#diff-c6f64a5895d9f518051cf83f8557f492ec6f5683a34c2914a46a894001e62726L364-R364)
> If you look at Maven Central at some point ( a few commits after this ) the artifact was renamed with the extra `_2.13`. This because `MillModule` depends on Scala where `MillPublishModule` depends on Java only.
> 
> **EDIT**: Actually it started to depend on `scalalib` so something happened in that PR that needs to be addressed.

We added Integration tests which obviously need mill. ;-) I now changed this dependency to be only a dependency of the actual test module, not the outer testng framework implementation.

lolgab commented on bsp/src/mill/bsp/Utils.scala:
```diff
@@ -56,24 +56,22 @@ object Utils {
           }
         )
-        classFingerprint.map(classF => classF._1.getName.stripSuffix("$"))
+        Seq.from(classFingerprint.map(classF => classF._1.getName.stripSuffix("$")))
```
Just curiosity: why not `.toSeq`?

lefou replied:
`.toSeq` gives a deprecation warning, and any fix is more verbose and results in the same call.

lolgab commented on scalalib/src/TestModule.scala:
```diff
@@ -0,0 +1,215 @@
+        "predefined TestModules: TestNg, Junit, Scalatest, ..."
+    if (frameworks.size != 1) {
+      Result.Failure(
+        "Since mill after-0.9.6 only one test framework per TestModule is supported. " ++ msg)
```
First time I see string concatenation with `++` :) Is it to not confuse it with the sum?

lefou replied:
I have no idea, haha. Will fix it...

lolgab reviewed (approved):
Good job on this one! Thank you!

sake92 commented:
Looks great really, seems like a great simplification!  
Just wondering if having `ivyDeps` for some frameworks and not for others is a bit inconsistent?  
Also, [MUnit](https://scalameta.org/munit/docs/getting-started.html#quick-start)? 😄

sake92 commented on scalalib/src/TestModule.scala:
```diff
@@ -0,0 +1,215 @@
+   * TestModule that uses Scalatest Framework to run tests.
+   * You need to provide the scalatest dependencies yourself.
+   */
+  trait Scalatest extends TestModule {
```
ScalaTest is more precise I think.

lefou commented:
> Looks great really, seems like a great simplification!
> Just wondering if having `ivyDeps` for some frameworks and not for others is a bit inconsistent?

It's mainly because those with come with `ivyDeps` don't provide a Sbt Test Interface. The side effect of course is, that they can be used without any addional `ivyDeps`.

> Also, [MUnit](https://scalameta.org/munit/docs/getting-started.html#quick-start)? smile

Oh, yeah. I never used it or sow it in action. But if you say, it works as smooth as all the others in mill, I can add a trait for it, too.

lefou commented:
Looks like latest CI runs fail because `org.tpolecat:tut-core` is no longer available. Probably an issue with bintray sunsetting?

sake92 commented:
Thx for adding MUnit, it's really awesome framework! :)  
Yeah, probably fails bcoz Bintray stuff. [Tut](https://github.com/tpolecat/tut) seems to be deprecated/abandoned anyways?

lefou commented:
> Looks like latest CI runs fail because `org.tpolecat:tut-core` is no longer available. Probably an issue with bintray sunsetting?

See #1269

lefou commented:
Thank you, @lolgab and @sake92, for you valuable feedback.
